crates/vm/wasm/src/cache.rs

```diff
@@ -32,18 +34,26 @@ impl Cache {
     where
         B: FnOnce() -> VmResult<(Module, Engine)>,
     {
-        // Cache hit - simply clone the module and return
-        if let Some(module) = self.inner.write_access().get(code_hash) {
+        // Pinned: Cache hit - simply clone the module and return
+        if let Some(module) = self.pinned.write_access().get(code_hash) {
+            return Ok(module.clone());
+        }
+        // Non pinned: Cache hit - simply clone the module and return
+        if let Some(module) = self.modules.write_access().get(code_hash) {
             return Ok(module.clone());
         }
 
         // Cache miss - build the module using the given builder method; insert
         // both the module and engine to the cache.
         let (module, engine) = builder()?;
-        self.inner
+        self.modules
             .write_access()
             .put(code_hash.clone(), (module.clone(), engine.clone()));
 
         Ok((module, engine))
     }
+
+    pub fn try_remove_from_modules(&self, code_hash: &Hash) {
+        self.modules.write_access().pop(code_hash);
+    }
```

Comment:
This should be unnecessary if you put `self.modules` and `self.pinned` under a single `Shared`.